### PR TITLE
Notification

### PR DESCRIPTION
**Pull request by Fralalala**
**UPDATED NEW ROUTES**

### - /notification/read/:id PUT
Takes in an id of a notification id and set its isRead to true

### - /notification/price POST
Creates a price notification for the product with the price change. This only handles price change. This route is only used
by a service

### - /notification/price/all GET
Gets all the notification of a user who's logged in.

### - /notification/price/unread GET
Gets all the unread notification of a user who's logged in

Key Notes: 

Notification Model has the following schema:

{
      id : int,                        // id of the Model
      type : string,               //type of notification assisted with the newly create enums.js file
      data: JSON,                //This is a JSON object that varies depending on the type of notification
      isRead: boolean,        //Boolean to refer to if the notification is read by the user
      user_id: id                  //the user who owns the notification
} 

data object properties for the type of "price"

{
        title: string,                            //Title of the product
        productId: int,                       //id of the product which  is the same id used to find it in the Product Model
        price: int,                               //the price that has been found by the service
        previousPrice: int,                 //the price of the product before the new price 

        listLocations: string [],          //the list of all UserList ids that the user owns where the product is kept. This is an array since 
                                                     //there's a possibility that a user will put the same product in two list. This can be used in the
                                                     //front end to say `You have ${listLocations.length} lists that contains the same product that had
                                                     //a price change`
}

### CODE REVIEW

File: server/routes/notification.js
Comment (johnsciutto) [inline server/routes/notification.js]: It would be cleaner to declare `promiseArr` as a constant (`const`), and then create a different array for looping over the `userList`. 
When you do what you did in line 64, the previous array still exists in memory, you're just reassigning the reference to something else. This is not a good idea. 
Look into using [`.map()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) instead of looping with a `for` loop and using `forEach` on the array below... it may give you exactly what you want 😉

File: server/routes/user.js
Comment (johnsciutto) [inline server/routes/user.js]: This could be another good situation for [`.map()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map).

File: server/routes/user.js
Comment (johnsciutto) [inline server/routes/user.js]: Same as above 😉

File: server/routes/user.js
Comment (johnsciutto) [inline server/routes/user.js]: The result of calling `NotificationModel.findOne({...})` and not finding a result is `null`?

File: server/routes/notification.js
Comment (LevoraShelby) [inline server/routes/notification.js]: Isn't user ID in `req.user`? Or has that changed?

File: server/routes/notification.js
Comment (LevoraShelby) [inline server/routes/notification.js]: This query can be combined with the one above with `await userNotification.update({isRead: true}, {where: notification_id, user_id: userId})`

File: server/routes/notification.js
Comment (LevoraShelby) [inline server/routes/notification.js]: Might be better as PUT

File: server/routes/notification.js
Comment (LevoraShelby) [inline server/routes/notification.js]: I might be misunderstanding, but I think this should be happening when a price drop is detected by a scraper service, and not by the user themselves. @sina-jamshidi, can you confirm?

File: server/routes/products.js
Comment (LevoraShelby) [inline server/routes/products.js]: This is something else I think ends up getting handled by the scraper service.

File: server/routes/user.js
Comment (LevoraShelby) [inline server/routes/user.js]: Associations could be used here to lower the number of queries needed in the request all the way down to one, although I've found them to be a bit iffy and hard to understand. If that fails, a raw query using a JOIN statement could still definitely be used. That aside though, I think what you have now works.

File: server/routes/user.js
Comment (LevoraShelby) [inline server/routes/user.js]: I think we should use the user ID in `req.user` for this route and the unread notifications route, otherwise anyone could view anyone else's notifications!

File: server/routes/notification.js
Comment (Fralalala) [reply server/routes/notification.js]: It is, I completely forgot about  the req.user. I'll update it using it

File: server/routes/notification.js
Comment (Fralalala) [reply server/routes/notification.js]: A notification will only happen on a price drop.

File: server/routes/products.js
Comment (Fralalala) [reply server/routes/products.js]: While development, I'm thinking that a product will be created once a user inputs a product link and that product link is deemed valid. This particular route is the one that I mainly used for testing

File: server/routes/notification.js
Comment (LevoraShelby) [reply server/routes/notification.js]: I understand. To clarify, I was wondering why there would specifically be a route for this, since I'm not sure this is functionality the user should be interfacing with.

File: server/routes/products.js
Comment (LevoraShelby) [reply server/routes/products.js]: Ah! Gotcha.

File: server/routes/notification.js
Comment (Fralalala) [reply server/routes/notification.js]: My knowledge might be lacking on this one lol. What would be a good way to make this inaccessible by a user and only by a scraping service?

Comment (LevoraShelby) [reply server/routes/notification.js]: Not sure on how are services are going to be implemented. That's probably going to be Sina's decision. If I were to guess we would probably set up scraping to happen at certain times of the day, but it's not something I think we would need a router for.

Comment (Fralalala) [reply server/routes/notification.js]: If I'm understanding this correctly, we wont be using a route for creating the model since it will be accessible by a user?

Comment (LevoraShelby) [reply server/routes/notification.js]: That would be my guess. Although the code you wrote in the route would still most likely get used.

Comment (sina-jamshidi) [reply server/routes/notification.js]: This is a good point. Notifications can be handled via service as well. This is okay for now, it can easily be refactored into a service if we need to by just copy pasting the code lol

File: server/models/productModel.js
Comment (sina-jamshidi) [inline server/models/productModel.js]: so a notification isn't actually tied to a product. 
The notification should have an id, title, image, description, and whether it is read. This way we can use it for many different things.

File: server/models/userNotificationsModel.js
Comment (sina-jamshidi) [inline server/models/userNotificationsModel.js]: 👍

File: server/models/productModel.js
Comment (Fralalala) [reply server/models/productModel.js]: A notification has a product, so a Notification Model would be "id" which is the id of the product and the "productId" which would reference the current information of the "productId". Now that I think about it, it seems like there would be a bug wherein they would all be referencing the same versions of the product.

Comment (sina-jamshidi) [reply server/models/productModel.js]: So now that I think about it, the reason the notification shouldn't link to a product id, is because say we have this situation:
1. We add `Laptop` to our list. Original name is `"Asus X"` and original price is `$20`
2. Price of `Laptop` drops to `$10`
3. We generate a notification telling user: `Laptop Asus X dropped from $20 to $10`
4. 2 days later Price of `Laptop` drops again to `$5` and name is changed to `"Asus Y"`
5. We generate a notification that tells the user: `Laptop Asus Y dropped from $10 to $5`

Now if we link the notification to a product id, and we are looking at our notification history, both of the notifications will say `Laptop named Asus Y dropped from $10 to $5`
So I think the notifications should have copies of this information. So they should just have `picture`, `type`, `message`
That way we can have different notifications like:
- "`Laptop named Asus Y dropped from $10 to $5`" with picture of the laptop and type: price change
- "User Emily has followed you" with profile picture of user and type: user update
- "You have received a message" with type: messages
etc.

File: server/routes/user.js
Comment (LevoraShelby) [inline server/routes/user.js]: This router is on the `/user` path, so I think it was fine as is. My `dbCamelCase` branch already reverts this since it's updating route names as well. I can change it back to `/get-user/:id` if you still think that would be better though.

File: server/routes/notification.js
Comment (LevoraShelby) [inline server/routes/notification.js]: Will this information be more useful than just the ID of the notification?

File: server/routes/notification.js
Comment (LevoraShelby) [inline server/routes/notification.js]: this has the potential to generate a lot of queries. You could do `await UserListModel.findAll({where: { id: listProds.map( (listProd) => listProd.id ) }})` just once. Also, I'm pretty sure this should be atomic with your last query. Try adding transactions if possible.

File: server/routes/notification.js
Comment (LevoraShelby) [inline server/routes/notification.js]: This bit is hard for me to follow. Could you add comments or otherwise make the code more readable?

**Review by LevoraShelby — COMMENTED** [review]
Just a couple things I might change, but otherwise looks good!
Also, if you could, I'd appreciate if you could wait until my `dbCamelCase` branch is merged into this one before merging into `dev`. Otherwise this branch will have to get merged into `dev` twice.

File: server/routes/notification.js
Comment (johnsciutto) [reply server/routes/notification.js]: I second that

File: server/routes/notification.js
Comment (Fralalala) [reply server/routes/notification.js]: I'm not sure what youre asking, but if youre asking about the returned value, I returned the notification with the updated isRead property. I figured returning the updated model would be better than just the default success value of the Model.update which is [1] .

File: server/routes/notification.js
Comment (sina-jamshidi) [inline server/routes/notification.js]: We should only have to query once. instead of querying the database to find the notification to update it, then querying the database again to return it.

File: server/routes/notification.js
Comment (sina-jamshidi) [inline server/routes/notification.js]: `const transaction = await sequelize.transaction();` No need to cut out the few letters. This will help readability

File: server/routes/notification.js
Comment (sina-jamshidi) [inline server/routes/notification.js]: `const newNotifications = {};`

File: server/routes/notification.js
Comment (sina-jamshidi) [inline server/routes/notification.js]: we don't need the backticks. `newNotifs[userList.user_id]` works

File: server/routes/notification.js
Comment (sina-jamshidi) [inline server/routes/notification.js]: a lot of unnecessarily shortened variable names

File: server/routes/notification.js
Comment (sina-jamshidi) [inline server/routes/notification.js]: I think you're looking for [bulkCreate](https://sequelize.org/master/class/lib/model.js~Model.html#static-method-bulkCreate)

File: server/routes/notification.js
Comment (sina-jamshidi) [inline server/routes/notification.js]: in this case I don't think we want to send all the notifications back. You can imagine that at some point this could be an array of thousands of notification objects and that's not necessary or efficient

File: server/routes/notification.js
Comment (sina-jamshidi) [inline server/routes/notification.js]: we're soon going to be making notifications for `follows` as well. Maybe this route and the next would make more sense to not be price specific and return all notification types. Maybe in the future we can add the ability to use a query to only get a certain type of notification

**Review by sina-jamshidi — CHANGES_REQUESTED** [review]

File: server/routes/notification.js
Comment (Fralalala) [reply server/routes/notification.js]: Will this process of using a model.save() be better or does it in effect does the same number of database interactions?

```
const notification = NotificationModel.findOne()

notification.isRead = true

const updatedNotification = await notification.save() 

res.send(updatedNotifications) 
```

File: server/routes/notification.js
Comment (Fralalala) [reply server/routes/notification.js]: Replaced sent data to an object msg.

`
{msg: "Success"} 
`

File: server/routes/notification.js
Comment (Fralalala) [reply server/routes/notification.js]: I could maybe just keep the code and make another route for fetching all notifications of the user.  This is just incase in the future wherein if we have a notifications page, users can filter out which notifications they want see

Comment (Fralalala) [reply server/routes/notification.js]: Added another route for getting user notifications with query properties for pagination.

File: server/routes/user.js
Comment (Fralalala) [reply server/routes/user.js]: I changed it back to just /:id. I initially added the /get-user to be much more specific and I had a case where if I had a route after the /:id route that is also the same route but different parameter name, it will use the /:id. If I'm not wrong, I dont think we wont be adding more routes in the user that just uses a parameter.

**Comment by LevoraShelby** [comment]
I noticed I get a 200 from `/notifications/read/:id` when attempting to read a notification that doesn't belong to the logged in user. It doesn't reflect in the db which is good, but maybe the user should get a 401 or 500 for something like that?

File: server/routes/notification.js
Comment (LevoraShelby) [reply server/routes/notification.js]: I _think_ you can add `returning: true` as an option when updating and it will return that new record. Take a look at https://sequelize.org/master/class/lib/model.js~Model.html#static-method-update

**Comment by Fralalala** [comment]
> I noticed I get a 200 from `/notifications/read/:id` when attempting to read a notification that doesn't belong to the logged in user. It doesn't reflect in the db which is good, but maybe the user should get a 401 or 500 for something like that?

Good catch, added validation for it

File: server/routes/notification.js
Comment (johnsciutto) [inline server/routes/notification.js]: Maybe this is something to talk with @sina-jamshidi , but I think it would make more sense to move the whole query validation/transformation logic to a separate function/middleware.

File: server/routes/notification.js
Comment (sina-jamshidi) [inline server/routes/notification.js:51]: nice extensible structure

File: server/routes/notification.js
Comment (sina-jamshidi) [reply server/routes/notification.js]: Yeah if you can modularize it a bit more even by moving them inside a function and returning an errorCode/message that would be good

**Review by sina-jamshidi — APPROVED** [review]
We're almost there! Just one more small note and you can merge!

**Pull request merged**
